tools/halide_image.h

```diff
@@ -1,278 +1,696 @@
-// This header defines a simple Image class which wraps a buffer_t. This is
-// useful when interacting with a statically-compiled Halide pipeline emitted by
-// Func::compile_to_file, when you do not want to link your processing program
-// against Halide.h/libHalide.a.
+// A simple Image class which wraps a buffer_t. This is useful
+// when interacting with a statically-compiled Halide pipeline emitted
+// by Func::compile_to_file, when you do not want to link your
+// processing program against Halide.h/libHalide.a.
 
-#ifndef HALIDE_TOOLS_IMAGE_H
-#define HALIDE_TOOLS_IMAGE_H
+#ifndef HALIDE_RUNTIME_IMAGE_H
+#define HALIDE_RUNTIME_IMAGE_H
 
-#include <cassert>
-#include <cstdlib>
-#include <limits>
 #include <memory>
 #include <vector>
-#include <stdint.h>  // <cstdint> requires C++11
+#include <cassert>
+#include <stdint.h>
+#include <string.h>
 
 #include "HalideRuntime.h"
 
 namespace Halide {
 namespace Tools {
 
-template<typename T>
-class Image {
-    struct Contents {
-        Contents(const buffer_t &b, uint8_t *a) : buf(b), ref_count(1), alloc(a) {}
-        buffer_t buf;
-        int ref_count;
-        uint8_t *alloc;
+template<typename Fn>
+void for_each_element(const buffer_t &buf, Fn f);
```

Comment:
"sample" implies to me that this is a sampled continuous signal, which is often not true. E.g. "sample" makes no sense for matrix entries. I went with something more generic to avoid that misdirection and to also match the name "elem_size".